### PR TITLE
HDF5: Fix Char Type Matching

### LINKED ISSUE #1431: Crash during file read on ppc64le due to incorrect datatype inference
**Describe the bug**
On `ppc64le`, reading an OpenPMD file crashes with : `Cannot change the datatype of a dataset.` This crash does not occur on any `linux-x86_64` systems or a `mac-aarch64` system.

The source code for the test is [here](https://github.com/fnalacceleratormodeling/synergia2/blob/devel3/src/synergia/bunch/tests/test_bunch.cc#L100-L119).

**Software Environment**
 - version of openPMD-api: `0.15.1`
 - installed openPMD-api via: `spack`
 - operating system: RHEL8
 - machine: `Wilson-cluster at Fermilab`
 - version of HDF5: `1.14.0`
 - version of ADIOS2: `2.8.3`
 - name and version of MPI: `Open MPI 5.1.0a1`
 
 **Additional Information**
 The cause of the crash is this throw: 
https://github.com/openPMD/openPMD-api/blob/32cb87d1b0012493287204d7a8c78af9a1141710/src/RecordComponent.cpp#L80-L81

And the variables within the function look like:
```
(gdb) print rc.m_dataset
$3 = {extent = {<std::_Vector_base<unsigned long, std::allocator<unsigned long> >> = {
      _M_impl = {<std::allocator<unsigned long>> = {<std::__new_allocator<unsigned long>> = {<No data fields>}, <No data fields>}, <std::_Vector_base<unsigned long, std::allocator<unsigned long> >::_Vector_impl_data> = {_M_start = 0x14185ff0, _M_finish = 0x14185ff8, _M_end_of_storage = 0x14185ff8}, <No data fields>}}, <No data fields>}, 
  dtype = openPMD::Datatype::CHAR, rank = 1 '\001', options = {
    _M_dataplus = {<std::allocator<char>> = {<std::__new_allocator<char>> = {<No data fields>}, <No data fields>}, _M_p = 0x1416a2c0 "{}"}, _M_string_length = 2, {
      _M_local_buf = "{}", '\000' <repeats 13 times>, _M_allocated_capacity = 32123}}}
(gdb) print d
$4 = {extent = {<std::_Vector_base<unsigned long, std::allocator<unsigned long> >> = {
      _M_impl = {<std::allocator<unsigned long>> = {<std::__new_allocator<unsigned long>> = {<No data fields>}, <No data fields>}, <std::_Vector_base<unsigned long, std::allocator<unsigned long> >::_Vector_impl_data> = {_M_start = 0x1417cbb0, _M_finish = 0x1417cbb8, _M_end_of_storage = 0x1417cbb8}, <No data fields>}}, <No data fields>}, 
  dtype = openPMD::Datatype::UCHAR, rank = 1 '\001', options = {
    _M_dataplus = {<std::allocator<char>> = {<std::__new_allocator<char>> = {<No data fields>}, <No data fields>}, _M_p = 0x7fffffff8d58 "{}"}, _M_string_length = 2, {
      _M_local_buf = "{}", '\000' <repeats 13 times>, _M_allocated_capacity = 32123}}}
``` 
The data types are indeed different `openPMD::Datatype::CHAR` vs `openPMD::Datatype::UCHAR`.

The type of the dataset (as inferred by `h5dump`) is `H5T_STD_U8LE`. But `openPMD` seems to interpret this to mean `openPMD::Datatype::CHAR`.

Within `synergia3`, the datatype is set to `unit8_t` for reading and writing. See `masks_datatype` in the `Bunch::read_openpmd_file`/`Bunch::write_openpmd_file` for reference [here](https://github.com/fnalacceleratormodeling/synergia2/blob/devel3/src/synergia/bunch/bunch.cc).

**Status with ADIOS2 backend**: No crash when using the `ADIOS2` backend.

### PR DESCRIPTION
**Pull request by ax3l**
In HDF5, there are only the signed and unsigned char type. The third `char` type is an alias to one or the other, different to the C/C++ fundamental types.

This tries to fix the type matching order to be platform independent between ppc64le/aarch64/x86-64: `signed char` stays a `signed char` now and the same for `unsigned char`. This also includes aliases, such as `std::uint8_t`/`std::int8_t` from `<cstdlib>`.

Due to the limited `char` aliasing in HDF5, these types are still mapped to platform-dependent types.

Additionally, this PR adds conversion for identical types as we also do for integer/float types.

Fix #1431

### CODE REVIEW

**Comment by franzpoeschel** [comment]
As discussed in #1431, the missing `char` datatype in HDF5 makes a full round-trip impossible for all char types without type casting. This test fails for HDF5 before and after the above change:

```cpp
namespace detail
{
template <typename Char>
void writeChar(Series &series, std::string const &component_name)
{
    auto component = series.iterations[0].meshes["E"][component_name];
    std::vector<Char> data(10);
    component.resetDataset({determineDatatype<Char>(), {10}});
    component.storeChunk(data, {0}, {10});
    series.flush();
}
template <typename Char>
void readChar(Series &series, std::string const &component_name)
{
    auto component = series.iterations[0].meshes["E"][component_name];
    std::vector<Char> data(10);
    auto chunk = component.loadChunk<Char>();
    series.flush();
    for (size_t i = 0; i < 10; ++i)
    {
        REQUIRE(data[i] == chunk.get()[i]);
    }
}
} // namespace detail

void char_roundtrip(std::string const &extension)
{
    Series write("../samples/char_rountrip." + extension, Access::CREATE);
    ::detail::writeChar<char>(write, "char");
    ::detail::writeChar<unsigned char>(write, "uchar");
    ::detail::writeChar<signed char>(write, "schar");
    write.close();

    Series read("../samples/char_rountrip." + extension, Access::READ_ONLY);
    ::detail::readChar<char>(read, "char");
    ::detail::readChar<unsigned char>(read, "uchar");
    ::detail::readChar<signed char>(read, "schar");
    read.close();
}
```

So, additionally I added a type-cast for equivalent char types.
As I used `switchType` for the `isSameChar` function template in `Datatype.hpp`, I needed to change the include structure a bit and remove the separation of `Datatype.hpp` and `DatatypeHelpers.hpp` which are now `Datatype.tpp`.

File: include/openPMD/Datatype.tpp [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+ * @tparam ReturnType The function template's return type.
+ * @tparam Action The struct containing the function template.
+ * @tparam Args The function template's argument types.
+ * @param dt The openPMD datatype.
+ * @param args The function template's arguments.
+ * @return Passes on the result of invoking the function template with the given
+ *     arguments and with the template parameter specified by dt.
+ */
+template <typename Action, typename... Args>
+constexpr auto switchType(Datatype dt, Args &&...args)
```
[the remark this answers is not in the document]
Comment (franzpoeschel) [reply include/openPMD/Datatype.tpp:101]: False positive, also this is an old function template moved to another header.

File: include/openPMD/Datatype.tpp [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+     *
+     * @tparam n As in switchType().
+     * @tparam ReturnType As in switchType().
+     * @tparam Action As in switchType().
+     * @tparam Args As in switchType().
+     */
+    template <int n, typename ReturnType, typename Action, typename... Args>
+    struct CallUndefinedDatatype
+    {
+        static ReturnType call(Args &&...args)
```
[the remark this answers is not in the document]
Comment (franzpoeschel) [reply include/openPMD/Datatype.tpp:69]: False positive, also this is an old function template moved to another header.

File: include/openPMD/Datatype.tpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (franzpoeschel) [reply include/openPMD/Datatype.tpp:101]: False positive, also this is an old function template moved to another header.

File: include/openPMD/Datatype.hpp [hunk 3]
```diff
@@ -651,6 +651,22 @@ inline bool isSameInteger(Datatype d)
         return false;
 }
 
+constexpr bool isChar(Datatype d)
```
Comment (ax3l) [inline include/openPMD/Datatype.hpp:661]: Please add Doxygen for `isChar` and `isSameChar` :)

File: include/openPMD/Datatype.tpp [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+ * and the GNU Lesser General Public License along with openPMD-api.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include "openPMD/Datatype.hpp"
+
+#include <string>
+#include <type_traits>
+#include <utility> // std::forward
```
Comment (ax3l) [inline include/openPMD/Datatype.tpp]:
```suggestion
#include <string>
#include <type_traits> // std::void_t
#include <utility> // std::forward

```

File: include/openPMD/Datatype.tpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ax3l) [reply include/openPMD/Datatype.tpp:101]: No really a false positive that it is a lot of code w/o inline docs, but due to the repetitive switch nature I think we clearly are good here :)

**Comment by s-sajid-ali** [comment]
Confirming that this fixes https://github.com/openPMD/openPMD-api/issues/1431 on `ppc64le`.

**Review by s-sajid-ali — APPROVED** [review]

**Pull request merged**